platforms/documentation/docs/src/docs/userguide/dep-man/basics/variant_aware_resolution.adoc

```diff
@@ -0,0 +1,142 @@
+
+1. `runtimeElements`, the regular variant offered by the `java-library` plugin
+2. `instrumentedJars`, the variant we have created
+
+== The consumer side
+
+=== 1. Add dependencies:
+
+First, on the consumer side, like any other project, we define the Java library as a dependency:
+
+====
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin",files="consumer/build.gradle.kts[tags=test_dependency]"]
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/groovy",files="consumer/build.gradle[tags=test_dependency]"]
+====
+
+At this point, Gradle will still select the default `runtimeElements` *variant* for your dependencies.
+This is because the `testRuntimeClasspath` configuration is requesting *artifacts* with the `jar` library elements *attribute*, while the producer defines the `instrumentedJars` *variant* with a different *attribute*.
+
+=== 2. Adjust the requested attributes:
+
+The `testRuntimeClasspath` configuration is modified to ask for "instrumented-jar" versions of the dependencies.
+This means that when Gradle resolves dependencies for this configuration, it will prefer JAR files that are marked as "instrumented":
+
+====
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin",files="consumer/build.gradle.kts[tags=ask-for-instrumented-classes]"]
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/groovy",files="consumer/build.gradle[tags=ask-for-instrumented-classes]"]
+====
+
+We can look at a report on the consumer side to view exactly what *attributes* of each dependency will be requested:
+
+[source,text]
+----
+$ .gradle resolvableConfigurations --configuration testRuntimeClasspath
+
+Attributes
+    - org.gradle.category            = library
+    - org.gradle.dependency.bundling = external
+    - org.gradle.jvm.version         = 11
+    - org.gradle.libraryelements     = instrumented-jar
+    - org.gradle.usage               = java-runtime
+----
+
+In this case, we see the `testRuntimeClasspath` configuration requests _instrumented classes_.
+
+=== 3. Handle missing variants with compatibility rules:
+
+Since not all dependencies (like `junit`) will provide instrumented classes, you need to define a compatibility rule that allows Gradle to fall back on the regular `jar` if an `instrumented-jar` is not available.
+
+To do this, we need to write a _compatibility rule_:
+
+====
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin",files="consumer/build.gradle.kts[tags=compatibility-rule]"]
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/groovy",files="consumer/build.gradle[tags=compatibility-rule]"]
+====
+
+Which we need to declare on the attributes schema:
```

Comment:
It might be worth pulling this up towards the top. We should first "define" the attribute on the attribute schema, then use that attribute on variants.

Reply:
Removing the compatibility rule since its too advanced.